Deploy Keycloak as an OAuth provider

KavanPrice opened the pull request:
This adds an OAuth mechanism to ACS with Kerberos integration.
- Defines "OpenID" deployment and service
- Defines IngressRoute for the service
- Creates a ConfigMap for a basic startup realm
  - Realm name: "factory_plus"
  - Realm clients:
    - grafana-oauth
  - Realm user federation:
    - Kerberos
- Adds the `auth.generic_oauth` section to `grafana.ini`

Only the client for the Grafana service is implemented here.

KavanPrice committed:
Add OpenID IngressRoute

KavanPrice committed:
Add Keycloak deployment

KavanPrice committed:
Add startup realm importing

KavanPrice committed:
Add oauth to Grafana config

KavanPrice committed:
Add keytab volume

amrc-benmorrow commented on deploy/templates/openid/realm-config.yaml:
```diff
@@ -0,0 +1,2740 @@
+            "subComponents": {},
+            "config": {
+              "serverPrincipal": [
+                "HTTP/openid.{{.Values.acs.baseUrl}}@{{.Values.acs.baseUrl | upper}}"
```
`.Values.identity.realm`. Realms don't necessarily follow the standard convention; in particular our v2 installation didn't.

djnewbould committed:
Add requested PR changes

Used local secret resource to remove secret lookup. Moved helm templating out of the values file.

djnewbould committed:
Add local secret grafana mount

KavanPrice committed:
Add realm creation to service-setup

KavanPrice committed:
Add service lookup for provided client configs

KavanPrice committed:
Add Grafana client config values

KavanPrice committed:
Fix keytab secret names

KavanPrice committed:
Add client_secret volume mount

KavanPrice committed:
Fix config and add startup backoff

KavanPrice committed:
Add generic client-secret fetching

KavanPrice committed:
Rename client-secret to be findable by service-setup

KavanPrice commented:
- Removed secret lookup and helm templating from Values
- Add service setup step for
  - Blank realm creation with Kerberos support
  - Dynamic client creation for clients defined in Values
- Remove `realm-config` monolith

amrc-benmorrow commented on acs-service-setup/lib/openid-realm.js:
```diff
@@ -0,0 +1,247 @@
+   */
+  async run() {
+    let base_realm = {
+      id: crypto.randomUUID(),
```
What happens the second time this runs? Will it try to create a new realm?

amrc-benmorrow replied:
OK. So we can't update anything, but that's OK for MVP. As long as service-setup finishes correctly in that case I think that's OK.

amrc-benmorrow commented on acs-service-setup/lib/openid-realm.js:
```diff
@@ -0,0 +1,247 @@
+            subComponents: {},
+            config: {
+              serverPrincipal: [
+                `HTTP/openid.${this.base_url}@${this.base_url.toUpperCase()}`,
```
We have `acs_config.realm` for this. The convention of using the uppercased domain is only a convention; our on-prem V2 installation didn't follow it.

amrc-benmorrow commented on acs-service-setup/lib/openid-realm.js:
```diff
@@ -0,0 +1,247 @@
+    }
+  }
+
+  async wait(milliseconds) {
```
`timers/promises` ?

KavanPrice replied:
Ah, I'm bad at JS - I didn't know this existed.

KavanPrice replied:
Fixed in [9a80e55](https://github.com/AMRC-FactoryPlus/amrc-connectivity-stack/pull/443/commits/9a80e558ed5885c071a1cc8675657c3be182c421)

amrc-benmorrow commented on deploy/templates/openid/openid.yaml:
```diff
@@ -0,0 +1,107 @@
+spec:
+  type: Random
+  principal: sv1openid@{{ .Values.identity.realm | required "values.identity.realm is required!" }}
+  secret: openid-keytabs/client
```
Do we need this? Is Keycloak a client of any F+ services?

KavanPrice replied:
Not yet. Removed in [15ed6a7](https://github.com/AMRC-FactoryPlus/amrc-connectivity-stack/pull/443/commits/15ed6a7419c46d520b9b72c8f1a528c895fd747b)

amrc-benmorrow commented on deploy/templates/openid/local-secrets.yaml:
```diff
@@ -0,0 +1,11 @@
+spec:
+  format: Password
+  secret: "keycloak-grafana-client"
+  key: "grafana"
```
This needs to generate multiple keys, one per client we are registering. For now I think we keep them all in the same Secret; this isn't ideal but makes the deployment much easier. So
* We need a `{{ range $k, $v := .Values.serviceSetup.config.openidClients }}` which generates multiple LocalSecret resources referencing different keys in the same Secret.
* The secret wants renaming to something more generic like `keycloak-clients`.
* The `extraSecretMounts` under the Grafana values needs an `items` property restricting Grafana to only read the `grafana` secret.

You could reasonably remove the `redirectHost` property and just use the key for the client service hostname.

KavanPrice replied:
Done in [e935fbb](https://github.com/AMRC-FactoryPlus/amrc-connectivity-stack/pull/443/commits/e935fbbca50a83d3ef34ebb757fd8b618a4450b8)

KavanPrice replied:
There was a bug here with accessing the namespace. Fixed in [e0b480f](https://github.com/AMRC-FactoryPlus/amrc-connectivity-stack/pull/443/commits/e0b480f6e13657b2a21dece9a3c5a05f310c2b36)

KavanPrice committed:
Fix misnamed grafana client ID

amrc-benmorrow commented on acs-service-setup/lib/openid-realm.js:
```diff
@@ -0,0 +1,247 @@
+      client_representation.redirectHost,
+    );
+    const content = await fs.readFile(secret_path, "utf8");
+    const client_secret = content.trim();
```
Is there any way to configure Keycloak to read the secret from its own filesystem at runtime, rather than reading the file contents here? This would be more reliable against the secret being updated. (In principle we should rotate all our secrets regularly.)

amrc-benmorrow replied:
A second container on the keycloak pod that called into the API to update the secrets every hour would also be a reasonable solution.

KavanPrice replied:
I'm looking at this. There are apparently ways of making it read a secret from a vault. Simplest solution might be, as you suggested, a container that just updates the secrets

KavanPrice committed:
Fix misnamed grafana secret

KavanPrice committed:
Change to use realm instead of base URL in ACS config

KavanPrice committed:
Use setTimeout from timers/promises

KavanPrice committed:
Timeout bugfix

KavanPrice committed:
Remove unused client keytab

KavanPrice committed:
Add dynamic LocalSecret generation

KavanPrice committed:
Fix helm loop bug

amrc-benmorrow committed:
Remove `items` from keytabs mount

This is left over from when we were pulling secrets from `krb5-keytabs`,
and isn't helpful here. It's causing the pod to fail to start because
the `client` key no longer exists.

amrc-benmorrow committed:
Allow OpenID errors to cause service-setup to exit

Without this we will silently fail when Keycloak is down.

amrc-benmorrow committed:
Reference client secrets correctly

Since changing the LocalSecret generation the keys have changed.

amrc-benmorrow committed:
Merge in testing/v4

amrc-benmorrow committed:
Remove Grafana Basic auth middleware

We should be using OAuth only now.

amrc-benmorrow committed:
Log realm details on creation

amrc-benmorrow committed:
Give Keycloack a PVC

We need to store the realm configuration somewhere.

amrc-benmorrow committed:
Log full client information

amrc-benmorrow committed:
Tell Grafana its external root URL

Otherwise OAuth doesn't work.

amrc-benmorrow committed:
Disable Grafana auth via the middleware

Rename 'Keycloak' to 'Factory+'. Users don't care about the
implementation.

amrc-benmorrow committed:
Grafana config changes

* `root_url` is under `server`.
* Disable the (useless) default login form.

amrc-benmorrow committed:
Java doesn't accept KRB5_CONFIG

There is a Java system property but I can't remember what it is.

amrc-benmorrow committed:
Keycloak needs to use Recreate strategy

We are backed with a ReadWriteOnce PVC.

KavanPrice committed:
Fix login and logout for basic users

Since we are using client_id for logout, Keycloak displays a logout confirmation page. This is skippable if we set id_token_hint. Grafana docs suggest Grafana will set this itself but it doesn't seem to work.

KavanPrice committed:
Add client role mapping

KavanPrice committed:
Add client role creation

KavanPrice committed:
Add admin-cli permissions to create users

KavanPrice committed:
Add default role to client

KavanPrice committed:
Add creation of admin user

KavanPrice committed:
Add client role mapping for admin user

amrc-benmorrow committed:
Refactor OpenID setup

The code was becoming very tangled and difficult to follow with
duplicated retry logic everywhere.

amrc-benmorrow committed:
Always create client roles

The list might have changed from last time. We don't attempt to delete
old roles at this point, this is for the user to handle if necessary.

amrc-benmorrow committed:
Use a LocalSecret for the admin service client

This is a client secret like any other; it needs to be preserved across
ACS upgrades, and so needs to come from a LocalSecret.

amrc-benmorrow committed:
Fix some logging

amrc-benmorrow committed:
Missing await

amrc-benmorrow committed:
We need to call try_fetch, not native fetch

amrc-benmorrow committed:
Error in admin user creation

amrc-benmorrow committed:
The admin@ account is no longer the service account

I'm getting weird 400 errors from this method and I wasn't sure if this was
the cause. But apparently not...

amrc-benmorrow committed:
Log if the admin user doesn't exist

amrc-benmorrow committed:
UserRepresentation.credentials is an array

This was giving a really unhelpful (and incorrect) 400 error.

amrc-benmorrow committed:
We must clone a Request to reuse it

Because a Request can contain a ReadableStream body we need to clone it
to reuse it.

amrc-benmorrow committed:
Rename the Keycloak bootstrap user

I am having a lot of trouble creating the admin@ user account; Keycloak
keeps returning 409 even though it later denies the account exists. I am
wondering if this is because of a conflict with the initial 'temporary'
bootstrap account username. Rename to avoid this; at the same time
change the password, this isn't linked to admin@ any more.

amrc-benmorrow committed:
Create admin@ linked to Kerberos

The Keycloak user-create endpoint is consistently returning 409 when we
attempt to create an account for the admin user. I think this may be
because we have federation set up to Kerberos and Keycloak refuses to
support non-Kerberos users in this realm. IMHO this is desirable in any
case.

We don't have a real human driving to perform the first-login flow, so
go round the houses a bit. First we request a token 'on behalf of'
admin@; this will fail because we haven't set up the profile, but will
create the user on the Keycloak side. Then we can go in with our user
management credentials and finish the setup.

amrc-benmorrow committed:
Assign admin user roles correctly

The `available` endpoint only returns roles which the user is not
already assigned to. We need to check existing assignments first.

amrc-benmorrow committed:
Remove Grafana admin user secret

This should now be configured via OpenID.

amrc-benmorrow committed:
We can't remove the grafana-internal admin user

If we don't specify an admin account secret Grafana generates one with
the username `admin`, which conflicts with the Kerberos account we are
trying to use via OpenID. Replace the Secret but change the username and
give it a password. It won't be possible to log in as this user in any
case as the login dialog is disabled.

amrc-benmorrow committed:
Assign Grafana roles from OAuth

amrc-benmorrow committed:
Try to get OpenID roles visible in Grafana

It seems like the only way to have the default install set up admin@ as
a Grafana admin is for Grafana to defer to Keycloak completely for role
information. This means that admin@ must also be able to manage the
roles assigned to other OpenID users; this information can no longer be
managed in Grafana.

* Allow assigning multiple default and admin roles.
* Allow assigning roles within builtin clients.
* Assign admin@ the roles needed to be a realm admin in Keycloak.
* Consistently use the key in openidClients as client ID.

amrc-benmorrow committed:
Restore OpenID role mapper

I'm guessing at this point; none of this appears to be documented. The
Grafana documentation doesn't mention needing this but it's not working
without, so try changing `claim.name` to the `roles` which I think
Grafana is expecting.

amrc-benmorrow committed:
We need to look up the client ID.

Confusingly, this is not the `clientId` field but the `id` field. This
is a generated UUID for `realm-management`.
